[#837] Multipart checkpoints

Kimahriman opened the pull request:
Resolves #837 

Adds a config that sets the maximum number of actions to include in a single file of a multipart checkpoint. If the total actions of a snapshot is larger than this, the checkpoint will be split up into multiple parts.

It is disabled by default so the behavior is purely opt-in, as most users probably don't know about actions or how to set this value. In the future it could be set to some sane default value if there's some consensus on what it should be.

vkorukanti commented:
Very useful feature and thanks for working on this.

> 
> * It looks like reading multipart checkpoints is already supported, is there anything that would need to be added on the read side?

As far as I know not needed anything on the reader side.

> * What test cases should be added?

We could start with simple tests which sets different values for `maxActionsPerFile` config and checks the output file generated. And finally the reader is able to read the file. 

> * Should it be a SQL conf? Or a table property? Or both options?

Seems like SQL config is better. We can have default value. No strong preference here.

> * If the setting is enabled, but the number of actions is under the threshold, should it create a multipart checkpoint with one part, or a single non-multipart checkpoint file?

I think creating single checkpoint file avoid the confusion.

Kimahriman commented:
Added a test for the multipart writing and updated to use a non-part file if only one file. Let me know what you think. Still need to figure out if there are any existing read path tests or if I need to make those as well

Kimahriman commented:
Working on updating the corrupt checkpoints tests to also test multipart

Kimahriman commented:
Updated the corrupt checkpoints tests to test for multipart too now, let me know how things look

Kimahriman commented:
Gentle ping, I think this will be helpful for the growing size of the transaction log with the upcoming file stats addition

vkorukanti commented on core/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:
```diff
@@ -290,18 +290,39 @@ object Checkpoints extends DeltaLogging {
+    }.getOrElse(1L)
+
+    val checkpointPaths = if (numCheckpointFiles > 1) {
+      checkpointFileWithParts(snapshot.path, snapshot.version, numCheckpointFiles.toInt)
```
just to be safe, add a check to `numCheckpointFiles.toInt` to throw exception if we exceeded the int limit.

Kimahriman replied:
any specific exception I should throw?

Kimahriman replied:
and looking at the scala api, is `isValidInt` a good way to test? or is just comparing to Int.MaxValue better?

vkorukanti commented on core/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:
```diff
@@ -370,38 +391,40 @@ object Checkpoints extends DeltaLogging {
+            throw DeltaErrors.failOnCheckpoint(src, dest)
+          }
+        } finally {
+          if (!renameDone) {
```
do we need to delete all the part files when an exception occurs?

Kimahriman replied:
yeah good call, let me think about that, probably just map the rename success and throw based on that instead

Kimahriman replied:
Hmm actually begs a bigger question. So in the case renames aren't used, we're not cleaning anything up. In non-multipart, that's theoretically fine because if it fails there's nothing to clean up. Should that attempt to clean itself up? And for rename case, should it try all renames, and then if any of them fail try to delete all existing files (both that succeeded and failed to rename)?

zsxwing replied:
It's better to clean up temp paths if we are using `rename`. These temp files won't be cleaned up by our automatic log cleanup.

Kimahriman replied:
So just cleanup all temp files for rename before throwing an exception and that's enough?

Kimahriman replied:
Updated so it waits to try to rename all the files (and cleanup failures) before throwing the exception

vkorukanti commented on core/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:
```diff
@@ -290,18 +290,39 @@ object Checkpoints extends DeltaLogging {
-    val checkpointRowCount = spark.sparkContext.longAccumulator("checkpointRowCount")
-    val numOfFiles = spark.sparkContext.longAccumulator("numOfFiles")
+    val sessionConf = spark.sessionState.conf
+    val maxActionsPerFile = sessionConf.getConf(DeltaSQLConf.DELTA_CHECKPOINT_MAX_ACTIONS_PER_FILE)
```
validate config value `maxActionsPerFile`?

Kimahriman replied:
Just add a validation to the conf itself? Just > 0 or any more validation you had in mind?

vkorukanti replied:
If we can add it to the conf that is fine. Yeah just a postive number check.

Kimahriman replied:
Add positive check

vkorukanti commented on core/src/test/scala/org/apache/spark/sql/delta/CheckpointsSuite.scala:
```diff
@@ -68,6 +72,93 @@ class CheckpointsSuite extends QueryTest
+
+        // 11 total actions, 7 new files
+        spark.range(30).repartition(7).write.format("delta").mode("append").save(path)
+        deltaLog.lastCheckpoint.foreach { checkpointMetadata =>
+          assert(checkpointMetadata.version == 2)
+          assert(checkpointMetadata.parts.isDefined)
+          assert(checkpointMetadata.parts.get == 2)
+        }
```
nit: these checks are repeated. May be an utility method will simplify the code?

Kimahriman replied:
Yeah i'll try to make one

Kimahriman replied:
Made helper for this

vkorukanti commented on core/src/test/scala/org/apache/spark/sql/delta/SnapshotManagementSuite.scala:
```diff
@@ -56,121 +68,208 @@ class SnapshotManagementSuite extends QueryTest with SQLTestUtils with SharedSpa
-        // Checkpoint 0 is corrupted. Verify that we can still create the snapshot using
-        // existing json files.
-        DeltaLog.forTable(spark, path).snapshot
+    for (useMultipart <- Seq(true, false)) {
+      // Test the first part and not the first part
+      val partsToCorrupt = if (useMultipart) {
+        Some(1) :: Some(2) :: Nil
+      } else {
+        None :: Nil
+      }
+      val maxActions = if (useMultipart) 1 else 10000
+      withSQLConf(DeltaSQLConf.DELTA_CHECKPOINT_MAX_ACTIONS_PER_FILE.key -> maxActions.toString) {
```
To minimize the changes in this file, can you try creating a test utility method

```
testWithAndWithoutMultipartCheckpoint(testname) { (isMultiParth, multiPartSize) =>
   // existing test code with some updates needed for 
}
```

```
def testWithAndWithoutMultipartCheckpoint(name)(f: (bool, int) -> Unit)) = {
   // here you can have the setup of multi-part or single part.
}
```

Kimahriman replied:
Yeah i'll try it

Kimahriman replied:
Done, let me know how it looks

vkorukanti reviewed (commented):
Apologies for taking so long to review. The PR looks good to me. Wondering if you had a chance to scale test this with a table having large number of actions?

Kimahriman commented:
> Apologies for taking so long to review. The PR looks good to me. Wondering if you had a chance to scale test this with a table having large number of actions?

I haven't but I can try. I don't foresee any issues since either way it's doing a repartition. Obviously if you set the max actions per file to 1 and have a million actions it might have performance issues, but that could be true for most of the settings

zsxwing commented on core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala:
```diff
@@ -408,6 +408,15 @@ trait DeltaSQLConfBase {
       .createWithDefault(true)
 
+  val DELTA_CHECKPOINT_MAX_ACTIONS_PER_FILE =
+    buildConf("checkpoint.maxActionsPerFile")
```
nit: since we call this `part`s in our code (such as checkpointFileWithParts/numCheckpointParts), I prefer to call this `checkpoint.partSize`.

Kimahriman replied:
changed the config name

zsxwing commented on core/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:
```diff
@@ -290,18 +290,39 @@ object Checkpoints extends DeltaLogging {
+    val sessionConf = spark.sessionState.conf
+    val maxActionsPerFile = sessionConf.getConf(DeltaSQLConf.DELTA_CHECKPOINT_MAX_ACTIONS_PER_FILE)
+
+    val res = snapshot.stateDF
+      .groupBy()
+      .agg(count("*").alias("rowCount"), count("add").alias("numOfFiles"))
+      .collect()
+      .head
```
This will trigger an extra Spark job to calculate the `rowCount`. Can we use an approximate value such as
```
val checkpointPartSize = ...
val estimatedNumParts = math.ceil(
        (snapshot.numOfFiles + snapshot.numOfRemoves).toDouble / checkpointPartSize
      ).toLong
```
to avoid launching a new Spark job?

Kimahriman replied:
Ah yeah good call

Kimahriman replied:
Updated to use just those values. Thought about always adding 2 for protocol + metadata to be more accurate but figured it didn't really matter since the files should dominate the overall count

zsxwing commented on core/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:
```diff
@@ -290,18 +290,39 @@ object Checkpoints extends DeltaLogging {
     // log store and decide whether to use rename.
     val useRename = deltaLog.store.isPartialWriteVisible(deltaLog.logPath, hadoopConf)
 
-    val checkpointRowCount = spark.sparkContext.longAccumulator("checkpointRowCount")
-    val numOfFiles = spark.sparkContext.longAccumulator("numOfFiles")
```
The reason we added these two accumulators is to verify the checkpoint we wrote is correct. We used to have a bug that we wrote an incomplete checkpoint, so we added these two accumulators for safety. If we compute them without writing the checkpointing, they would be meaningless. Could you keep them?

Kimahriman replied:
Yeah I'll add them back since I'm getting rid of the agg now anyway

Kimahriman replied:
Added back in

zsxwing commented on core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala:
```diff
@@ -408,6 +408,15 @@ trait DeltaSQLConfBase {
+  val DELTA_CHECKPOINT_MAX_ACTIONS_PER_FILE =
+    buildConf("checkpoint.maxActionsPerFile")
+      .internal()
+      .doc("Maximum number of actions to add to a checkpoint file before splitting it up. " +
```
```
The limit at which we will start parallelizing the checkpoint. We will attempt to 
write maximum of this many actions per checkpoint.
```

Kimahriman replied:
Updated

zsxwing reviewed (commented):
Thanks for the update! Only a few minor comments.

Kimahriman commented on core/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:
```diff
@@ -370,26 +391,27 @@ object Checkpoints extends DeltaLogging {
       }
+      // Fail the checkpoint write after letting any failed renames cleanup
+      failedPaths.headOption.foreach { case (src, dest) =>
+        // There should be only one writer writing the checkpoint file, so there must be
```
Ah yeah whoops missed this

Kimahriman commented on core/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:
```diff
@@ -370,26 +391,27 @@ object Checkpoints extends DeltaLogging {
+        val dest = new Path(paths(index))
+        val fs = dest.getFileSystem(hadoopConf)
+        var renameDone = false
+        if (!fs.rename(src, dest)) {
```
Wouldn't that immediately throw the exception after the first rename fails (and the finally block is executed), so any remaining files to rename would be left there? I wasn't sure how to handle that case, just catch any error and ignore it but just throw the `failOnCheckpoint` if anything goes wrong?

zsxwing replied:
Ah, you are right. I think we can just delete all `src`s in `finally` if `rename` is not completed, such as the following 

```
var renameDone = false
try {
  rename written files
  renameDone = true
} finally {
  if (!renameDone) {
     writtenPaths.foreach { writtenPath =>
          scala.util.Try { fs.delete(writtenPath, false) }
     }
  }
}
```

Kimahriman replied:
ok took a stab at this, wasn't sure how to handle rename fails vs rename throws an error

Kimahriman commented on core/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:
```diff
@@ -40,7 +40,7 @@ import org.apache.spark.sql.{Column, DataFrame, SparkSession}
 import org.apache.spark.sql.catalyst.expressions.{Cast, ElementAt, Literal}
 import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
-import org.apache.spark.sql.functions.{col, struct, when}
+import org.apache.spark.sql.functions.{col, count, struct, when}
```
thanks, I forget the scalastyle doesn't pick up unused things

Kimahriman commented on core/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:
```diff
@@ -292,16 +292,36 @@ object Checkpoints extends DeltaLogging {
+    val paths = checkpointPaths.map(_.toString)
     val base = snapshot.stateDS
-      .repartition(1)
+      .repartition(paths.length)
```
Sounds good

Kimahriman replied:
updated

zsxwing commented:
@Kimahriman could you also resolve the conflicts?

Kimahriman commented:
Rebased

zsxwing commented:
> Rebased

Thanks! By the way, you don't have to rebase. We usually just use `git merge` because it's easier to deal with conflicts (solve the conflicts once). Our merge script will handle it.

zsxwing commented on core/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:
```diff
@@ -388,24 +409,29 @@ object Checkpoints extends DeltaLogging {
+        writtenPaths.zipWithIndex.foreach { case (writtenPath, index) =>
+          val src = new Path(writtenPath)
+          val dest = new Path(paths(index))
+          val fs = dest.getFileSystem(hadoopConf)
```
one last minor comment (sorry. I missed it): We can use `val fs = snapshot.path.getFileSystem(hadoopConf)` outside and then reuse the `fs` object for `rename` and `delete`. This would be faster when people disable file system cache.

Kimahriman replied:
good point, done

Kimahriman commented:
> > Rebased
> 
> Thanks! By the way, you don't have to rebase. We usually just use `git merge` because it's easier to deal with conflicts (solve the conflicts once). Our merge script will handle it.

Hah I guess you're right, I've just gotten so use to rebasing open source stuff

zsxwing reviewed (approved):
LGTM. Thanks!
